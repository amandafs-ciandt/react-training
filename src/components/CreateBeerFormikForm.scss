.beer-form {
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  margin: 0 3rem;

  &__input {
    width: 100%;
    display: grid;
    grid-template-columns: 13% 85%;
    column-gap: 2%;
    align-items: center;
    justify-items: flex-end;
    margin-bottom: 1.5rem;

    &.checkbox-input {
      justify-items: flex-start;
    }

    .ant-select {
      width: 100%;
    }
  }

  &__label {
    height: 2rem;
    color: rgba(0, 0, 0, 0.85);
    font-size: 14px;

    &.required-field::before {
      display: inline-block;
      margin-right: 4px;
      color: #ff4d4f;
      font-size: 14px;
      font-family: SimSun, sans-serif;
      line-height: 1;
      content: '*';
    }
  }

  Button {
    margin-left: 15%;
    width: fit-content;
  }
}
