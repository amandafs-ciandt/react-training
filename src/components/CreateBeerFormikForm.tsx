import React from 'react';
import { submitForm } from '../effector/event';
import { useFormik } from 'formik';
import * as yup from 'yup';

import { Input, Checkbox, Button, Select } from 'antd';
import './CreateBeerFormikForm.scss';

const { Option } = Select;

const CreateBeerFormikForm = ({ notify }) => {
  const beerSchema = yup.object().shape({
    beerName: yup.string().required('Required'),
    beerType: yup.string().required('Required'),
    ingredients: yup.string().required('Required'),
  });

  const beerForm = useFormik({
    initialValues: {
      beerName: '',
      beerType: '',
      hasCorn: false,
      ingredients: '',
    },
    validationSchema: beerSchema,
    onSubmit: (values, { resetForm }) => {
      submitForm(values);

      notify('Form successfully submitted!');
      resetForm();
    },
  });

  return (
    <form onSubmit={beerForm.handleSubmit} className='beer-form'>
      <div className='beer-form__input'>
        <label htmlFor='beerName' className='beer-form__label required-field'>
          Beer name:
        </label>
        <Input
          name='beerName'
          value={beerForm.values.beerName}
          onChange={beerForm.handleChange}
        />
      </div>

      <div className='beer-form__input'>
        <label htmlFor='beerType' className='beer-form__label required-field'>
          Beer type:
        </label>
        <Select
          placeholder='Select a beer type'
          value={beerForm.values.beerType}
          onChange={(value) => beerForm.setFieldValue('beerType', value)}>
          <Option value='ale'>Ale</Option>
          <Option value='lager'>Lager</Option>
          <Option value='malt'>Malt</Option>
          <Option value='stout'>Stout</Option>
        </Select>
      </div>

      <div className='beer-form__input checkbox-input'>
        <label></label>
        <Checkbox
          checked={beerForm.values.hasCorn}
          onChange={() =>
            beerForm.setFieldValue('hasCorn', !beerForm.values.hasCorn)
          }>
          Has corn
        </Checkbox>
      </div>

      <div className='beer-form__input'>
        <label
          htmlFor='ingredients'
          className='beer-form__label required-field'>
          Ingredients:
        </label>
        <Input.TextArea
          name='ingredients'
          value={beerForm.values.ingredients}
          onChange={beerForm.handleChange}
        />
      </div>

      <Button
        type='primary'
        htmlType='submit'
        disabled={!(beerForm.isValid && beerForm.dirty)}>
        Submit
      </Button>
    </form>
  );
};

export default CreateBeerFormikForm;
