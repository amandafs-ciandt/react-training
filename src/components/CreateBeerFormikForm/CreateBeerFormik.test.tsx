import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import sinon from "sinon";

import CreateBeerFormikForm from ".";

const onNotifyMock = jest.fn();

const setUp = () => {
  return shallow(<CreateBeerFormikForm notify={onNotifyMock} />);
};

const fillFormValues = (component: ShallowWrapper) => {
  component.find("Input").simulate("change", {
    persist: () => {},
    target: {
      name: "beerName",
      value: "Budweiser",
    },
  });

  component.find("Select").simulate("change", {
    persist: () => {},
    target: {
      name: "beerType",
      value: "lager",
    },
  });

  component.find("Checkbox").simulate("change", {
    persist: () => {},
    target: {
      name: "hasCorn",
      checked: true,
    },
  });

  component.find("TextArea").simulate("change", {
    persist: () => {},
    target: {
      name: "ingredients",
      value: "Water, barley malt, rice, yeast and hops",
    },
  });
};

describe("CreateBeerFormikForm", () => {
  let createBeerFormikFormComponent: ShallowWrapper;

  beforeEach(() => {
    createBeerFormikFormComponent = setUp();
  });

  it("should work", () => {
    expect(createBeerFormikFormComponent).toMatchSnapshot();
  });

  it("should render form correctly", () => {
    const [
      inputLength,
      selectLength,
      checkboxLength,
      textareaLength,
      totalFieldsLength,
    ] = [1, 1, 1, 1, 4];

    expect(createBeerFormikFormComponent.find("Input").length).toBe(
      inputLength
    );
    expect(createBeerFormikFormComponent.find("Select").length).toBe(
      selectLength
    );
    expect(createBeerFormikFormComponent.find("Checkbox").length).toBe(
      checkboxLength
    );
    expect(createBeerFormikFormComponent.find("TextArea").length).toBe(
      textareaLength
    );
    expect(createBeerFormikFormComponent.find(".beer-form__input").length).toBe(
      totalFieldsLength
    );
  });

  it("should show the correct beer types on select", () => {
    const beerTypes = ["ale", "lager", "malt", "stout"];
    const optionElements = createBeerFormikFormComponent.find("Option");

    expect(optionElements.length).toBe(beerTypes.length);

    expect(optionElements.get(0).props.value).toEqual(beerTypes[0]);
    expect(optionElements.get(1).props.value).toEqual(beerTypes[1]);
    expect(optionElements.get(2).props.value).toEqual(beerTypes[2]);
    expect(optionElements.get(3).props.value).toEqual(beerTypes[3]);
  });

  it("should update beerName when it is changed", () => {
    createBeerFormikFormComponent.find("Input").simulate("change", {
      persist: () => {},
      target: {
        name: "beerName",
        value: "Budweiser",
      },
    });

    const newValue = createBeerFormikFormComponent.find("Input").props().value;

    expect(newValue).toEqual("Budweiser");
  });

  it("should update beerType when it is changed", () => {
    createBeerFormikFormComponent.find("Select").simulate("change", {
      persist: () => {},
      target: {
        name: "beerType",
        value: "lager",
      },
    });

    // to do: improve this
    const newValue = createBeerFormikFormComponent.find("Select").props()
      .value as any;

    expect(newValue.target.value).toEqual("lager");
  });

  it("should update hasCorn when it is changed", () => {
    createBeerFormikFormComponent.find("Checkbox").simulate("change", {
      persist: () => {},
      target: {
        name: "hasCorn",
        checked: true,
      },
    });

    const newValue = createBeerFormikFormComponent.find("Checkbox").props()
      .checked;

    expect(newValue).toEqual(true);
  });

  it("should update ingredients when it is changed", () => {
    const ingredientsValue = "Water, barley malt, rice, yeast and hops";

    createBeerFormikFormComponent.find("TextArea").simulate("change", {
      persist: () => {},
      target: {
        name: "ingredients",
        value: ingredientsValue,
      },
    });

    const newValue = createBeerFormikFormComponent.find("TextArea").props()
      .value;

    expect(newValue).toEqual(ingredientsValue);
  });

  it("should have submit button disabled at first", () => {
    const submitButton = createBeerFormikFormComponent.find("Button");

    expect(submitButton.props().disabled).toBeTruthy();
  });

  it("should have submit button enabled when form is valid", () => {
    fillFormValues(createBeerFormikFormComponent);

    const submitButton = createBeerFormikFormComponent.find("Button");

    expect(submitButton.props().disabled).toBeFalsy();
  });
});
