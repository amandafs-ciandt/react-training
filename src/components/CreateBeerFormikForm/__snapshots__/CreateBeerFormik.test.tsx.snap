// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateBeerFormikForm should work 1`] = `ShallowWrapper {}`;
